### PR TITLE
Get rid of AppConstants in `application/palemoon`

### PR DESCRIPTION
**Pull request by JustOff**
`AppConstants.jsm` is used in the following two places in `application/palemoon`:

* shell component
* autocomplete.xml

This PR uses preprocessing to get rid of it.

Tag #65.

### CODE REVIEW

File: application/palemoon/components/shell/ShellService.jsm [hunk 1]
```diff
@@ -26,20 +25,18 @@ var ShellServiceInternal = {
...
-
-    if (AppConstants.platform == "linux") {
-      if (this.shellService) {
-        let linuxShellService = this.shellService
-                                    .QueryInterface(Ci.nsIGNOMEShellService);
-        return linuxShellService.canSetDesktopBackground;
-      }
+#if defined(XP_WIN) || defined(XP_MACOSX)
+    return true;
+#elif XP_LINUX
```
Comment (adisib) [inline application/palemoon/components/shell/ShellService.jsm]: Shouldn't this be "#elif defined(XP_LINUX)" ?

Comment (mattatobin) [reply application/palemoon/components/shell/ShellService.jsm]: No.. one does not have to do that for one codition.

Comment (adisib) [reply application/palemoon/components/shell/ShellService.jsm]: I meant in terms of consistency with how it is done elsewhere in the codebase. But I guess that is just a personal nitpick and not worth bringing up.

**Comment by JustOff** [comment]
Please put this on hold for a while, I have to re-check something.

**Comment by JustOff** [comment]
Could someone tell me why Mozilla uses such a complex method to detect `android` and `linux` platforms?

https://github.com/MoonchildProductions/UXP/blob/ac25827a87d86f1cf9e48aab6605f77a2c89041a/toolkit/modules/AppConstants.jsm#L147-L162

Why the detection uses `MOZ_WIDGET_GTK`, `MOZ_WIDGET_ANDROID` and `XP_LINUX`, instead of just check
```
#elif ANDROID
  "android",
#elif XP_LINUX
  "linux",
```
Do I also need to build a complex rule or `#if defined(XP_LINUX) && !defined(ANDROID)` is enough?

**Comment by wolfbeast** [comment]
Android and MacOSX are both Linux variants that have `XP_LINUX` set. the `ANDROID` define is only set in nss, and the most definitive way to check for specific OS is to check the widget decision.
Look at the conditionals - it needs that order. If you have GTK widgets, it's always a variant of pure Linux. if XP_MACOSX is set, you're on the Mac flavor, if you have Android widgets, you're on Android. If you pass all those checks and XP_LINUX is still set, then it's a non-GTK flavor of pure Linux. You can't simplify this without breaking things.

**Comment by JustOff** [comment]
Well, thank you for the clarification. But in this case, is not it better to continue using AppConstants, at least for platform detection?

**Comment by JustOff** [comment]
One more question, is it possible that `MOZ_WIDGET_GTK` is defined, while `XP_LINUX` is not?

**Comment by JustOff** [comment]
> You're not here to optimize appConstants, you're here to do away with them ;)

Sorry, but I can't work with the code not trying to understand and optimize it. And in this case the problem is aggravated by the fact that if you literally follow the `AppConstants` rules for detecting the platform, you will get an ugly code that does not occur anywhere else in the sources.

> Android and MacOSX are both Linux variants that have XP_LINUX set.

I think you were mistaken, they are variants of `XP_UNIX`, but not `XP_LINUX`, see [init.configure#L527-L570](https://github.com/MoonchildProductions/UXP/blob/36f73c8cd27e62cbd3e85939d6fe11a240e3416f/build/moz.configure/init.configure#L527-L570).

So, the question of whether it is possible that `MOZ_WIDGET_GTK` is defined, while `XP_LINUX` is not remains very relevant in order to move on. Perhaps @mattatobin or @trav90 can add something about this?

**Comment by wolfbeast** [comment]
It does mean, however, that if you see an appconstants check with "linux" in it, that you need to check what it was before in Mozilla code, because it could be either of `MOZ_WIDGET_GTK` or `XP_LINUX` that you have to change it to.

**Comment by wolfbeast** [comment]
Also, why are you even looking at appconstants.jsm if all you are doing here is removing the use of them from a specific part of the tree? Unless you intend to remove specific appconstant values from the entire tree, you should leave that module alone.

**Comment by mattatobin** [comment]
Here is a simple resolution. No one touches AppConstants but me. Stick to what ever form is used in the file at hand in the codebase and I will work through converting it back to preprocessing over time.

**Comment by wolfbeast** [comment]
I'll close this PR then

**Comment by JustOff** [comment]
Ok, deal.

**Pull request closed without merging**